service/src/java/org/apache/hive/service/server/HiveServer2.java

```diff
@@ -1201,63 +1204,63 @@ private static void startHiveServer2() throws Throwable {
+      Map<String, Integer> customPools = CompactorUtil.getPoolConf(hiveConf);
 
       StringBuilder sb = new StringBuilder(2048);
       sb.append("This HS2 instance will act as Compactor with the following worker pool configuration:\n");
-      sb.append("Global pool size: ").append(numWorkers).append("\n");
-
-      LOG.info("Initializing the compaction pools with using the global worker limit: {} ", numWorkers);
-      while (numWorkers > 0 && entries.size() > 0) {
-        Map.Entry<String, String> entry = entries.remove(0);
-        String poolName = entry.getValue();
-        int poolWorkers = hiveConf.getInt(entry.getKey(), 0);
+      sb.append("Global pool size: ").append(numWorkers.get()).append("\n");
 
+      LOG.info("Initializing the compaction pools with using the global worker limit: {} ", numWorkers.get());
+      customPools.forEach((poolName, poolWorkers) -> {
         if (poolWorkers == 0) {
           LOG.warn("Compaction pool ({}) configured with zero workers. Skipping pool initialization", poolName);
           sb.append("Pool not initialized, 0 size: ").append(poolName).append("\n");
-          continue;
         }
-        if (poolWorkers > numWorkers) {
-          LOG.warn("Global worker pool exhausted, compaction pool ({}) will be configured with less workers than the " +
-              "required number. ({} -> {})", poolName, poolWorkers, numWorkers);
-          poolWorkers = numWorkers;
+        else if (numWorkers.get() == 0) {
+          LOG.warn("There are no available workers for the following compaction pool: {} ", poolName);
+          sb.append("Pool not initialized, no remaining free workers: ").append(poolName).append("\n");
         }
+        else {
+          if (poolWorkers > numWorkers.get()) {
+            LOG.warn("Global worker pool exhausted, compaction pool ({}) will be configured with less workers than the " +
+                "required number. ({} -> {})", poolName, poolWorkers, numWorkers.get());
+            poolWorkers = numWorkers.get();
+          }
 
-        LOG.info("Initializing compaction pool ({}) with {} workers.", poolName, poolWorkers);
-        for (int i = 0; i < poolWorkers; i++) {
-          Worker w = new Worker();
-          w.setPoolName(poolName);
-          CompactorThread.initializeAndStartThread(w, hiveConf);
-          sb.append("Worker - Name: ").append(w.getName()).append(", Pool: ").append(poolName)
-              .append(", Priority: ").append(w.getPriority()).append("\n");
+          LOG.info("Initializing compaction pool ({}) with {} workers.", poolName, poolWorkers);
+          IntStream.range(0, poolWorkers).forEach(i -> {
+            Worker w = new Worker();
+            w.setPoolName(poolName);
+            CompactorThread.initializeAndStartThread(w, hiveConf);
+            startedWorkers.put(poolName, startedWorkers.getOrDefault(poolName, 0)+1);
+            sb.append("Worker - Name: ").append(w.getName()).append(", Pool: ").append(poolName)
```

Comment:
not sure, but wouldn't this be more readable
````
sb.append(
String.format("Worker - Name: %s, Pool: %s, Priority: %i, w.getName(), poolName,w.getPriority() ));
````

Reply:
This was in the existing code, but fixed as you suggested.